Add support for impersonation for certificate requests

paulwilljones opened the pull request:
Resolves #331

howardjohn commented on pkg/server/auth.go:
```diff
@@ -79,16 +98,72 @@ func (s *Server) authRequest(ctx context.Context, csrPEM []byte) (string, bool)
+			return identities, false
+		}
+	} else if !identitiesMatch([]string{s.impersonatedIdentity}, csr.URIs) {
+		log.Error(fmt.Errorf("%v != %v", s.impersonatedIdentity, csr.URIs), "failed to match URIs with identities")
```
```suggestion
		log.Error(fmt.Errorf("%v != %v", s.impersonatedIdentity, csr.URIs), "failed to match URIs with impersonated identities")
```

nit: might be nice to help distinguish

howardjohn reviewed (commented):
:+1:

inteon commented on pkg/server/node_auth.go:
```diff
@@ -0,0 +1,111 @@
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// ClusterNodeAuthorizer is a component that implements a subset of Kubernetes Node Authorization
```
@paulwilljones does Istio have a similar ClusterNodeAuthorizer in their own implementation?

paulwilljones replied:
Yes, [`ClusterNodeAuthorizer`](https://github.com/istio/istio/blob/1.22.1/security/pkg/server/ca/node_auth.go#L74) is the same, however, I've reimplemented it here to make sure the pod informer is synced before the index is created.

inteon replied:
Could you add a link to that implementation in this file?

SgtCoDFish commented:
/ok-to-test

inteon commented:
@paulwilljones the automated tests found some issues: https://storage.googleapis.com/cert-manager-prow-artifacts/pr-logs/pull/cert-manager_istio-csr/336/pull-cert-manager-istio-csr-verify/1813889849741021184/build-log.txt

paulwilljones commented:
/retest

paulwilljones commented:
@inteon @SgtCoDFish I added an e2e test for Ambient. Can we add this to Prow for Istio 1.22 onwards?

inteon commented:
> @inteon @SgtCoDFish I added an e2e test for Ambient. Can we add this to Prow for Istio 1.22 onwards?

The tests are defined here: https://github.com/cert-manager/testing/blob/master/config/jobs/cert-manager/istio-csr/cert-manager-istio-csr-presubmits.yaml
We probably should start by adding this as an optional test so other PRs are not blocked.

paulwilljones commented:
> > @inteon @SgtCoDFish I added an e2e test for Ambient. Can we add this to Prow for Istio 1.22 onwards?
> 
> The tests are defined here: https://github.com/cert-manager/testing/blob/master/config/jobs/cert-manager/istio-csr/cert-manager-istio-csr-presubmits.yaml We probably should start by adding this as an optional test so other PRs are not blocked.

Added https://github.com/cert-manager/testing/pull/1050 to run Ambient e2e tests for 1.22

SgtCoDFish commented on pkg/server/node_auth.go:
```diff
@@ -0,0 +1,130 @@
+		return fmt.Errorf("caller (%v) is not allowed to impersonate", caller)
+	}
+
+	// // Next, make sure the identity they want to impersonate is valid, in general
```
turbo nitpick (not a blocker): There are a few comments here that have `// //`, which I guess is an artifact from copy-pasting?

SgtCoDFish commented on pkg/server/node_auth.go:
```diff
@@ -0,0 +1,130 @@
+		return err
+	}
+
+	// // Finally, we validate the requested identity is running on the same node the caller is on
```
here too!

SgtCoDFish commented on test/e2e/suite/runtimeconfiguration/runtimeconfiguration.go:
```diff
@@ -81,6 +91,10 @@ var _ = framework.CasesDescribe("runtime configuration", func() {
 	})
 
 	It("it should use the runtime configured issuer", func() {
+		//TODO: paulwilljones fix for Ambient
```
question: are you intending to do this before merging this PR?

paulwilljones replied:
this currently fails for ambient as the `istioctl proxy-config` command to retrieve the certificates is not valid for ztunnel, and `istioctl experimental ztunnel-config certificates` doesn't return the required output. We can block the PR on this until there's a workaround or fix once `istioctl` moves out of experimental.

SgtCoDFish reviewed (changes requested):
With no exaggeration this is one of the best PRs I've been able to review! Thank you!

I've got a few little bits here and there and a few questions about how to proceed with this, but I don't see any major blockers!

SgtCoDFish commented:
It looks like #357 caused quite a few conflicts with this PR - sorry about that!

paulwilljones commented on go.mod:
```diff
@@ -18,8 +18,8 @@ require (
-	istio.io/api v1.22.3
-	istio.io/istio v0.0.0-20240418224740-ed90e14d3473
+	istio.io/api v1.23.0-alpha.0.0.20240717200518-ef9163fb1200
+	istio.io/istio v0.0.0-20240723082231-2fc58e8b9b6d
```
as discussed, `go mod tidy` sets these versions and compatibility between `istio.io/api` and `istio.io/istio` is difficult to correlate.

inteon commented on deploy/charts/istio-csr/values.yaml:
```diff
@@ -205,6 +205,8 @@ app:
       # NB: This variable is named incorrectly; it controls private key algorithm, not signature algorithm.
       signatureAlgorithm: "RSA"
+    # A list of service accounts that are allowed to use node authentication for CSRs, eg. istio-system/ztunnel
+    caTrustedNodeAccounts: ""
```
should we default this to `istio-system/ztunnel`?

inteon replied:
in which situation do we want to change this value?

howardjohn replied:
Users may run ztunnel under a different service account. For example https://istio.io/latest/docs/ambient/install/platform-prerequisites/#google-kubernetes-engine-gke

inteon commented:
I'll upgrade the istio package in https://github.com/cert-manager/istio-csr/pull/385, this should reduce the diff for this PR a bit.

SgtCoDFish commented on deploy/charts/istio-csr/values.yaml:
```diff
@@ -205,6 +205,8 @@ app:
       # The type of private key to generate for the serving certificate. Only RSA (default) and ECDSA are supported.
       # NB: This variable is named incorrectly; it controls private key algorithm, not signature algorithm.
       signatureAlgorithm: "RSA"
+    # A list of service accounts that are allowed to use node authentication for CSRs, eg. istio-system/ztunnel
+    caTrustedNodeAccounts: ""
```
nitpick (non-blocking): It might be worth saying that this is a comma-separated list, to make it easier for people to know how to specify the values.

```suggestion
    # A comma-separated list of service accounts that are allowed to use node authentication for CSRs, eg. "istio-system/ztunnel"
    caTrustedNodeAccounts: ""
```

SgtCoDFish commented on make/test-e2e.mk:
```diff
@@ -72,23 +72,32 @@ endif
 .PHONY: e2e-setup-istio
 e2e-setup-istio: | kind-cluster install $(NEEDS_KUBECTL) $(bin_dir)/scratch/istioctl-$(ISTIO_VERSION)
-	$(bin_dir)/scratch/istioctl-$(ISTIO_VERSION) install -y -f ./make/config/istio/istio-config-$(ISTIO_VERSION).yaml
+	$(bin_dir)/scratch/istioctl-$(ISTIO_VERSION) install -y -f ./make/config/istio/istio-config-$(ISTIO_VERSION).yaml $(ISTIO_INSTALL_OPIONS)
```
suggestion: There's a typo in `ISTIO_INSTALL_OPIONS` - should be `ISTIO_INSTALL_OPTIONS`.

SgtCoDFish commented on test/e2e/suite/mtls/mtls.go:
```diff
@@ -54,6 +54,18 @@ var _ = framework.CasesDescribe("mTLS correctness", func() {
 			},
 		}
 
+		injectLabel = func() map[string]string {
+			if f.Config().AmbientEnabled {
+				return map[string]string{
+					"istio.io/dataplane-mode": "ambient",
+				}
+			} else {
+				return map[string]string{
+					"istio-injection": "enabled",
+				}
+			}
+		}()
```
question: Maybe an obvious question (and not really related to the PR), but just checking I've understood right: I guess there's no situation where you'd set dataplane-mode to ambient and enable injection? I can't imagine why that would be desirable, just want to check.

SgtCoDFish replied:
Thanks, that's really helpful 👍

SgtCoDFish reviewed (changes requested):
Got a few comments on this but nothing huge. Feels like it's really close!

SgtCoDFish reviewed (approved):
/lgtm
/approve

Thanks for all your work on this, it's really cool and I'm hoping we'll get to release this out for people to play with soon 🚀
